src/array_api_typing/_array.py

```diff
@@ -67,10 +74,131 @@ def dtype(self, /) -> DTypeT_co:
         ...
 
 
+class HasDevice(Protocol):
+    """Protocol for array classes that have a device attribute."""
+
+    @property
+    def device(self) -> object:  # TODO: more specific type
+        """Hardware device the array data resides on."""
+        ...
+
+
+class HasMatrixTranspose(Protocol):
+    """Protocol for array classes that have a matrix transpose attribute."""
+
+    @property
+    def mT(self) -> Self:  # noqa: N802
+        """Transpose of a matrix (or a stack of matrices).
+
+        If an array instance has fewer than two dimensions, an error should be
+        raised.
+
+        Returns:
+            Self: array whose last two dimensions (axes) are permuted in reverse
+                order relative to original array (i.e., for an array instance
+                having shape `(..., M, N)`, the returned array must have shape
+                `(..., N, M))`.  The returned array must have the same data type
+                as the original array.
+
+        """
+        ...
+
+
+class HasNDim(Protocol):
```

Comment:
Is there a situation where you'd wanna use `HasNDim` over e.g. `HasShape`? Otherwise we probably should keep this private, given that 

> There should be one-- and preferably only one --obvious way to do it.
> Although that way may not be obvious at first unless you're Dutch.

Reply:
> We do need a stub library... Maybe that's the easier thing to do first.

Stub libraries only apply to, well, libraries. Stub libraries also can't be overridden or something, so libraries that have other methods besides the one from the array-api won't be able to use it. So I don't see how that would work

Reply:
@jorenham can we get in protocols for the methods? 

https://github.com/data-apis/array-api-typing/issues/22#issuecomment-3215249910

Reply:
Methods and attributes of Array.